source/language/openpulse.rst

```diff
@@ -0,0 +1,316 @@
+such as local-oscillator frequencies, which do not vary throughout program
+execution. Again it is expected that vendors of quantum hardware provide a
+method for manipulating those static settings if appropriate.
+
+There are two kinds of channels: transmit channels (sending input to a quantum
+device) and receive channels (reading output from a quantum device).
+
+The ``channel`` type is a compile-time type representing a channel.
+
+A value of type ``channel`` is retreived with the special functions
+``transmit`` and ``receive``. These functions take an optional string argument
+followed by 1 or more physical qubits.
+
+The hardware vendor, during compilation, will substitute valid calls to
+``transmit`` and ``receive`` with their corresponding channels.
+
+A recommended convention: calling ``transmit`` with a single qubit argument
+should return the channel for driving single qubit gates, and calling
+``receive`` with a single qubit argument should return the channel for reading
+out the state of that qubit.
+
+.. code-block:: none
+
+   // Get the default input channel for qubit 0
+   transmit(%0)
+   // Get the default output channel for qubit 1
+   receive(%1)
+   // Get an input channel named "measure" for qubit 1
+   // Could be used to specify the measurement stimulus channel
+   transmit("measure", %1)
+   // Get an output channel named "cr1" for qubit 0
+   receive("cr1", %0)
+   // Get the default output channel for qubits 0 and 1
+   receive(%0, %1)
+
+Frames
+------
+
+It turns out to be quite useful to keep track of a set of carrier signals
+throughout the execution of a program. These carrier signals are called "frames"
+and are defined by :math:$A*e^{i\left(2\pi f t + \theta\right)}$, where `f` is
+frequency, `theta` is phase, and `A` is a scaling factor.
+
+Frames are attached to channels by a 1:1 correspondence. Every frame must 
+specify a frequency before any pulses can be played on that channel. The special 
+``prelude`` defcal is a good place to place these instructions.
+
+.. code-block:: none
+
+   defcal prelude {
+      // Required: set the frequency to 5GHz
+      set_frequency transmit(%1), 5e9;
+      // Optional: set the starting phase (otherwise it would start at 0)
+      set_phase transmit(%1), pi;
+   }
+
+Frames track the set of (frequency, phase, scale) throughout the program and 
+are manipulated using the following `set_` and `shift_` commands. Then when
+a pulse is played the carrier signal of the frame is mixed with the pulse itself
+to create the final signal which will be applied on the channel.
+
+Frequency has no default since it must be set at the beginning. Phase defaults
+to 0 and is a value of type ``angle``. Scale defaults to 1.0 and is a value of
+type ``float``. The exact precision of these three parameters is hardware
+specific.
+
+frameof command
+~~~~~~~~~~~~~~~
+
+The ``frame`` type is a compile-time type represnting a frame. It is useful
```

Comment:
If a frame can be later assigned to a channel I think this resolves the discussion above.